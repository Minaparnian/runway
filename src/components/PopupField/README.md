```js
const ButtonContent = (
  <div>
    <div>Open</div>
    <div>Popup</div>
  </div>
);

<PopupField buttonContent={ButtonContent}>
  {({ closePopup }) => <Button onClick={closePopup} label="Close Popup" />}
</PopupField>;
```
