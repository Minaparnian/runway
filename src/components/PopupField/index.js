import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Transition } from 'react-transition-group';

class PopupField extends Component {
  state = {
    open: false
  };

  onEntered = () => {
    document.addEventListener('click', this.handleClickOutside);
    this.closeButton.focus();
  };

  onExited = () => {
    document.removeEventListener('click', this.handleClickOutside);
    this.props.onClose();
    this.fieldButton.focus();
  };

  handleClickOutside = event => {
    if (this.wrapper && !this.wrapper.contains(event.target)) {
      this.togglePopup();
    }
  };

  togglePopup = () => {
    this.setState(({ open }) => ({
      open: !open
    }));
  };

  render() {
    const {
      openAriaLabel,
      closeAriaLabel,
      popupAriaLabel,
      navigationAriaLabel,
      children,
      onBlur,
      buttonContent
    } = this.props;

    const { open } = this.state;

    const content =
      typeof children === 'function'
        ? children({
            closePopup: this.togglePopup
          })
        : children;

    const transitionStyles = {
      entering: {
        transform: 'translateY(100vh)'
      },
      entered: {
        transform: 'translateY(0vh)'
      },
      exiting: {
        transform: 'translateY(100vh)'
      }
    };

    return (
      <div
        ref={el => {
          this.wrapper = el;
        }}
      >
        <button
          aria-haspopup="dialog"
          aria-expanded={open}
          aria-label={`${open ? closeAriaLabel : openAriaLabel}`}
          onClick={this.togglePopup}
          onBlur={onBlur}
          ref={el => {
            this.fieldButton = el;
          }}
          type="button"
        >
          {buttonContent}
        </button>
        <Transition
          in={open}
          onExited={this.onExited}
          onEntered={this.onEntered}
          timeout={300}
          unmountOnExit
        >
          {state => (
            <div
              css={{
                background: 'white',
                boxSizing: 'border-box',
                height: '100%',
                left: 0,
                padding: '15px',
                position: 'fixed',
                top: 0,
                transform: 'translateY(0vh)',
                transition: `transform 300ms ease-in-out`,
                width: '100%',
                zIndex: 1000,
                ...transitionStyles[state]
              }}
              role="dialog"
              aria-label={popupAriaLabel}
            >
              <button
                aria-label={`${closeAriaLabel}. ${navigationAriaLabel}`}
                onClick={this.togglePopup}
                type="button"
                ref={el => {
                  this.closeButton = el;
                }}
              >
                {'<'}
              </button>
              <div>{content}</div>
            </div>
          )}
        </Transition>
      </div>
    );
  }
}

PopupField.propTypes = {
  buttonContent: PropTypes.oneOfType([PropTypes.string, PropTypes.element])
    .isRequired,
  openAriaLabel: PropTypes.string,
  closeAriaLabel: PropTypes.string,
  popupAriaLabel: PropTypes.string,
  navigationAriaLabel: PropTypes.string,
  onClose: PropTypes.func,
  onBlur: PropTypes.func,
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.func])
};

PopupField.defaultProps = {
  openAriaLabel: 'Open dialog',
  closeAriaLabel: 'Close dialog',
  popupAriaLabel: 'Dialog is open',
  navigationAriaLabel: 'Tab to navigate',
  onClose: () => {},
  onBlur: () => {},
  children: null
};

export default PopupField;
