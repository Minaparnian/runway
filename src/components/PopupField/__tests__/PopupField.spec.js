import React from 'react';
import { mount } from 'enzyme';
import PopupField from '..';

function openPopup(component) {
  component.instance().setState({ open: true });
}

describe('PopupField', () => {
  let component;

  it('renders correctly with defaults', () => {
    component = mount(<PopupField buttonContent="Popup Field" />);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with props', () => {
    beforeAll(() => {
      component = mount(
        <PopupField
          buttonContent="Popup Field"
          openAriaLabel="Open Aria Label"
          closeAriaLabel="Close Aria Label"
          popupAriaLabel="Popup Aria Label"
          navigationAriaLabel="Tab to navigate"
          onClose={e => e}
        />
      );
      openPopup(component);
    });

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with jsx button content', () => {
    beforeAll(() => {
      const ButtonContent = <div>Button</div>;
      component = mount(
        <PopupField
          buttonContent={ButtonContent}
          openAriaLabel="Open Aria Label"
          closeAriaLabel="Close Aria Label"
          popupAriaLabel="Popup Aria Label"
          navigationAriaLabel="Tab to navigate"
          onClose={e => e}
        />
      );
      openPopup(component);
    });

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with one child element', () => {
    component = mount(
      <PopupField buttonContent="Popup Field">
        <span>Child Element</span>
      </PopupField>
    );
    openPopup(component);

    expect(component).toMatchSnapshot();
  });

  it('renders correctly with one child function', () => {
    component = mount(
      <PopupField buttonContent="Popup Field">
        {({ closePopup }) => <button type="button" onClick={closePopup} />}
      </PopupField>
    );
    component.instance().setState({ open: true });

    expect(component).toMatchSnapshot();
  });

  describe('onEntered()', () => {
    beforeEach(() => {
      document.addEventListener = jest.fn();
      component = mount(
        <PopupField
          buttonContent="Popup Field"
          openAriaLabel="Open Aria Label"
          closeAriaLabel="Close Aria Label"
          popupAriaLabel="Popup Aria Label"
          navigationAriaLabel="Tab to navigate"
          onClose={e => e}
        />
      );
      openPopup(component);
      jest.spyOn(component.instance().closeButton, 'focus');
      component.instance().onEntered();
    });

    afterEach(() => {
      document.addEventListener.mockReset();
      component.instance().closeButton.focus.mockReset();
    });

    it('adds click event listener', () => {
      const addEventListenerCalls = document.addEventListener.mock.calls;

      expect(addEventListenerCalls.length).toBe(1);
      expect(addEventListenerCalls[0][0]).toBe('click');
      expect(addEventListenerCalls[0][1]).toBe(
        component.instance().handleClickOutside
      );
    });

    it('focuses the close button', () => {
      expect(component.instance().closeButton.focus).toHaveBeenCalled();
    });
  });

  describe('onExited()', () => {
    const mockOnClose = jest.fn();

    beforeEach(() => {
      document.removeEventListener = jest.fn();
      component = mount(
        <PopupField
          buttonContent="Popup Field"
          openAriaLabel="Open Aria Label"
          closeAriaLabel="Close Aria Label"
          popupAriaLabel="Popup Aria Label"
          navigationAriaLabel="Tab to navigate"
          onClose={mockOnClose}
        />
      );
      jest.spyOn(component.instance().fieldButton, 'focus');
      component.instance().onExited();
    });

    afterEach(() => {
      document.removeEventListener.mockReset();
      mockOnClose.mockReset();
      component.instance().fieldButton.focus.mockReset();
    });

    it('removes click event listener', () => {
      const removeEventListenerCalls = document.removeEventListener.mock.calls;

      expect(removeEventListenerCalls.length).toBe(1);
      expect(removeEventListenerCalls[0][0]).toBe('click');
      expect(removeEventListenerCalls[0][1]).toBe(
        component.instance().handleClickOutside
      );
    });

    it('calls this.props.onClose()', () => {
      expect(mockOnClose.mock.calls.length).toBe(1);
    });

    it('focuses the field button', () => {
      expect(component.instance().fieldButton.focus).toHaveBeenCalled();
    });
  });

  describe('handleClickOutside()', () => {
    beforeEach(() => {
      component = mount(
        <PopupField
          buttonContent="Popup Field"
          openAriaLabel="Open Aria Label"
          closeAriaLabel="Close Aria Label"
          popupAriaLabel="Popup Aria Label"
          navigationAriaLabel="Tab to navigate"
          onClose={e => e}
        />
      );
      openPopup(component);
      jest.spyOn(component.instance(), 'togglePopup');
    });

    afterEach(() => {
      component.instance().togglePopup.mockReset();
    });

    it('calls togglePopup() if the click was outside', () => {
      const componentInstance = component.instance();

      componentInstance.handleClickOutside({ target: document });
      expect(componentInstance.togglePopup).toHaveBeenCalled();
    });

    it('does nothing if the click was inside', () => {
      const componentInstance = component.instance();

      componentInstance.handleClickOutside({
        target: componentInstance.wrapper
      });
      expect(componentInstance.togglePopup).not.toHaveBeenCalled();
    });
  });

  describe('togglePopup()', () => {
    beforeEach(() => {
      component = mount(
        <PopupField
          buttonContent="Popup Field"
          openAriaLabel="Open Aria Label"
          closeAriaLabel="Close Aria Label"
          popupAriaLabel="Popup Aria Label"
          navigationAriaLabel="Tab to navigate"
          onClose={e => e}
        />
      );
    });

    it('opens the popup', () => {
      component.instance().togglePopup();

      expect(component.state('open')).toBe(true);
    });

    it('closes the popup', () => {
      const componentInstance = component.instance();

      componentInstance.setState({ open: true });
      componentInstance.togglePopup();

      expect(component.state('open')).toBe(false);
    });
  });
});
