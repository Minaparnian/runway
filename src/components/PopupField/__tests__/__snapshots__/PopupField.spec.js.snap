// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupField renders correctly with defaults 1`] = `
<PopupField
  buttonContent="Popup Field"
  closeAriaLabel="Close dialog"
  navigationAriaLabel="Tab to navigate"
  onBlur={[Function]}
  onClose={[Function]}
  openAriaLabel="Open dialog"
  popupAriaLabel="Dialog is open"
>
  <div>
    <button
      aria-expanded={false}
      aria-haspopup="dialog"
      aria-label="Open dialog"
      onBlur={[Function]}
      onClick={[Function]}
      type="button"
    >
      Popup Field
    </button>
    <Transition
      appear={false}
      enter={true}
      exit={true}
      in={false}
      mountOnEnter={false}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      timeout={300}
      unmountOnExit={true}
    />
  </div>
</PopupField>
`;

exports[`PopupField renders correctly with jsx button content 1`] = `
<PopupField
  buttonContent="Popup Field"
  closeAriaLabel="Close dialog"
  navigationAriaLabel="Tab to navigate"
  onBlur={[Function]}
  onClose={[Function]}
  openAriaLabel="Open dialog"
  popupAriaLabel="Dialog is open"
>
  <div>
    <button
      aria-expanded={false}
      aria-haspopup="dialog"
      aria-label="Open dialog"
      onBlur={[Function]}
      onClick={[Function]}
      type="button"
    >
      Popup Field
    </button>
    <Transition
      appear={false}
      enter={true}
      exit={true}
      in={false}
      mountOnEnter={false}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      timeout={300}
      unmountOnExit={true}
    />
  </div>
</PopupField>
`;

exports[`PopupField renders correctly with one child element 1`] = `
<PopupField
  buttonContent="Popup Field"
  closeAriaLabel="Close dialog"
  navigationAriaLabel="Tab to navigate"
  onBlur={[Function]}
  onClose={[Function]}
  openAriaLabel="Open dialog"
  popupAriaLabel="Dialog is open"
>
  <div>
    <button
      aria-expanded={false}
      aria-haspopup="dialog"
      aria-label="Open dialog"
      onBlur={[Function]}
      onClick={[Function]}
      type="button"
    >
      Popup Field
    </button>
    <Transition
      appear={false}
      enter={true}
      exit={true}
      in={false}
      mountOnEnter={false}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      timeout={300}
      unmountOnExit={true}
    />
  </div>
</PopupField>
`;

exports[`PopupField renders correctly with one child function 1`] = `
<PopupField
  buttonContent="Popup Field"
  closeAriaLabel="Close dialog"
  navigationAriaLabel="Tab to navigate"
  onBlur={[Function]}
  onClose={[Function]}
  openAriaLabel="Open dialog"
  popupAriaLabel="Dialog is open"
>
  <div>
    <button
      aria-expanded={false}
      aria-haspopup="dialog"
      aria-label="Open dialog"
      onBlur={[Function]}
      onClick={[Function]}
      type="button"
    >
      Popup Field
    </button>
    <Transition
      appear={false}
      enter={true}
      exit={true}
      in={false}
      mountOnEnter={false}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      timeout={300}
      unmountOnExit={true}
    />
  </div>
</PopupField>
`;

exports[`PopupField renders correctly with props 1`] = `
<PopupField
  buttonContent="Popup Field"
  closeAriaLabel="Close dialog"
  navigationAriaLabel="Tab to navigate"
  onBlur={[Function]}
  onClose={[Function]}
  openAriaLabel="Open dialog"
  popupAriaLabel="Dialog is open"
>
  <div>
    <button
      aria-expanded={false}
      aria-haspopup="dialog"
      aria-label="Open dialog"
      onBlur={[Function]}
      onClick={[Function]}
      type="button"
    >
      Popup Field
    </button>
    <Transition
      appear={false}
      enter={true}
      exit={true}
      in={false}
      mountOnEnter={false}
      onEnter={[Function]}
      onEntered={[Function]}
      onEntering={[Function]}
      onExit={[Function]}
      onExited={[Function]}
      onExiting={[Function]}
      timeout={300}
      unmountOnExit={true}
    />
  </div>
</PopupField>
`;
